pdns/recursordist/lwres.cc

```diff
@@ -623,6 +841,41 @@ static LWResult::Result asyncresolve(const ComboAddress& address, const DNSName&
           }
         }
       }
+      if (g_cookies && !*chained) {
+        auto [done, result] = incomingCookie(log, address, localip, *now, cookieSentOut, edo, doTCP, *lwr, cookieFoundInReply);
+        if (done) {
+          return result;
+        }
+      }
+    }
+
+    // Case: we sent out a cookie but did not get one back
+    if (cookieSentOut && !cookieFoundInReply && !*chained) {
+      ++t_Counters.at(rec::Counter::cookieNotInReply);
+      auto lock = s_cookiestore.lock();
+      auto found = lock->find(address);
+      if (found != lock->end()) {
+        switch (found->getSupport()) {
+        case CookieEntry::Support::Probing:
+          VLOG(log, "No cookie in reply from " << address.toString() << ", was probing, setting support to Unsupported" << endl);
+          found->setSupport(CookieEntry::Support::Unsupported, now->tv_sec);
+          ++t_Counters.at(rec::Counter::cookieProbeUnsupported);
+          break;
+        case CookieEntry::Support::Unsupported:
+          // We could have detected the server does not support cookies in the meantime
+          VLOG(log, "No cookie in reply from " << address.toString() << ", cookie state is Unsupported, fine" << endl);
+          break;
+        case CookieEntry::Support::Supported:
+          // RFC says: ignore replies not containing any cookie info, equivalent to timeout
+          VLOG(log, "No cookie in reply from " << address.toString() << ", cookie state is Supported, dropping packet as if it timed out)" << endl);
+          return LWResult::Result::Timeout;
+          break;
+        }
+      }
+      else {
```

Comment:
Does this mean that when I prune or delete cookies for an IP, my first query to that auth is *always* treated as a timeout?

Reply:
Prune is unlikely, as the state is updated on sending out the cookie.
For delete: yes: if you delete in between send and receive.